partner integration docs v0 (#6262)
ClickHouse/clickhouse-docs: docs/integrations/integration-development/testing-your-integration.md

---
slug: /integrations/integration-development/testing-your-integration
sidebar_label: 'Testing your integration'
sidebar_position: 3
title: 'Testing your ClickHouse integration'
description: 'Entry-level validation matrix for integrations on ClickHouse Cloud and self-hosted open source.'
keywords: ['partner', 'integration', 'testing', 'validation', 'example datasets', 'ClickHouse Cloud', 'open source']
doc_type: 'guide'
---

# Testing your ClickHouse integration

> Validate your integration against both ClickHouse deployment modes and datasets that exercise ClickHouse's type system at meaningful scale before you submit it for review. This page defines what "tested" means at the entry level. Formal validation is a separate process for partners progressing to higher partnership tiers.

Reviewer: ```suggestion
Validate your integration against both ClickHouse deployment modes and datasets that exercise ClickHouse's type system at a meaningful scale before you submit it for review. This page defines what "tested" means at the entry level. Formal validation is a separate process for partners progressing to higher partnership tiers.
```

~"at a meaningful scale" -> could we link to some guidance here?~
Ignore, mentioned below

See [Building integrations](/integrations/integration-development/building-integrations) for ingestion and consumption paths, and [Documenting your integration](/integrations/integration-development/documenting-your-integration) for how to publish your results.

## Test matrix {#test-matrix}

Cover both deployment modes. Most customers run one or the other, and behavior differs in places (auth, networking, available features).

- **ClickHouse Cloud:** sign up for a [free trial](https://clickhouse.com/cloud). No credit card is required for the development tier
- **Self-hosted (open source):** use the latest stable release from [GitHub releases](https://github.com/ClickHouse/ClickHouse/releases). The [install guide](/install) is the fastest path to a local instance with Docker

Test against both, and document any feature gaps in your integration page.

## What to test {#what-to-test}

**Functional correctness.** Exercise every code path your integration exposes: ingestion, querying, schema discovery, error handling, and reconnection. If your product surfaces SQL to end users, confirm that the queries your UI generates round-trip cleanly.

> **Type-system coverage.** ClickHouse supports arrays, tuples, maps, JSON, nested, LowCardinality, Decimal, Date and DateTime variants, UUID, IPv4 and IPv6, enums, and aggregate-function types. Integrations often hit issues with nested arrays, deeply nested tuples, and JSON columns. Your client library and UI should handle these gracefully. At minimum, fail with a readable error instead of silently truncating or misrendering.

Reviewer: ```suggestion
**Type-system coverage.** ClickHouse supports arrays, tuples, maps, JSON, nested, LowCardinality, Decimal, Date and DateTime variants, UUID, IPv4 and IPv6, enums, and aggregate-function types. Integrations often hit issues with nested arrays, deeply nested tuples, and JSON columns. Your client library and UI should handle these gracefully, and at a minimum, fail with a readable error instead of silently truncating or misrendering.
```

**Scale.** Test at result-set sizes and row counts your customers will run. For user-facing BI, that often means tables with hundreds of millions to billions of rows, and result sets from single aggregates to tens of thousands of rows. Unbounded reads (`SELECT *`) should fail predictably or paginate, not hang.

**Authentication.** Validate at least one TLS-enabled connection. If you expose auth configuration, test every mode you document (username and password over TLS, mTLS, SSL client certificate).

**Connection lifecycle.** Confirm sane behavior on dropped connections, server restarts, and slow queries. Many escalations trace back to connection handling rather than query semantics.

## Recommended example datasets {#recommended-example-datasets}

> The full set is at [Example datasets](/getting-started/example-datasets). These four cover most integration testing needs:

Reviewer: ```suggestion
The full set can be found in the [**Example datasets**](/getting-started/example-datasets) section. The following four datasets cover most integration testing needs:
```
